### PR TITLE
GW different cost ranks

### LINKED ISSUE #112: Specific ranks for each cost in `GW`

### PR DESCRIPTION
**Pull request by michalk8**
In this PR:
- allow for passing different ranks/tolerances for geometries in GW when converting them to LR
- quadratic solver no longer modifies the problem in-place
- add `QuadraticProblem.to_low_rank(seed: int = 0)` method (similar could be done for `LinearProblem in the future`)

closes #112

### CODE REVIEW

File: ott/core/gromov_wasserstein.py [hunk 1]
```diff
@@ -137,52 +137,11 @@ def update(
...
-            prob.geom_xy, pointcloud.PointCloud
-        ) and prob.geom_xy.is_squared_euclidean:
-          prob.geom_xy = prob.geom_xy.to_LRCGeometry(prob.fused_penalty)
-        else:
-          prob.geom_xy = prob.geom_xy.to_LRCGeometry(
-              rank=self.cost_rank, tol=self.cost_tol
-          )
+    # consider converting problem first if using low-rank solver
+    if self.is_low_rank and prob._should_convert_to_low_rank:
+      prob = prob.to_low_rank()
```
Comment (marcocuturi) [inline ott/core/gromov_wasserstein.py:144]: this makes sense, and it's indeed cleaner, thanks

File: ott/core/quad_problems.py [hunk 2]
```diff
@@ -121,6 +121,14 @@ class QuadraticProblem:
     gw_unbalanced_correction: True (default) if the unbalanced version of
       Sejourne et al. (Neurips 2021) is used, False if tau_a and tau_b
       only affect the inner Sinhkorn loop.
+    ranks: Ranks of the cost matrices, see
+      :meth:`~ott.geometry.geometry.Geometry.to_LRCGeometry`. Used when
+      geometries are *not* :class:`~ott.geometry.pointcloud.PointCloud` with
+      `'sqeucl'` cost function. If `-1`, the geometries will not be converted
+      to low-rank. If :class:`int`, rank shared across all geometries.
```
Comment (marcocuturi) [inline ott/core/quad_problems.py:128]: If :class:`int`, geometries provided as inputs will be converted to low-rank geometries using ...

maybe discuss the case where we have a Tuple and the order with which it is used in that case?

**Review by marcocuturi — APPROVED** [review]
Thanks Michal, small comment in doc, could be easier to understand what tuple refers to.

File: ott/core/quad_problems.py (hunk 2 above)
Comment (michalk8) [reply ott/core/quad_problems.py]: Done!

**Pull request merged**